### PR TITLE
Fix exception raised on empty pattern in search plugin in preferences

### LINKED ISSUE #6106: `ValueError: WindowsPath('.') has an empty name` raised when interacting with the file extension readers preference line edit
## 🐛 Bug

A `ValueError` exception raises when interacting with the file extension patterns line edit at least on Windows (`Preferences > Plugins > File extension readers > Start typing filename pattern... line edit`):

pref_error

## To Reproduce

1. Open preferences dialog
2. Go to the `Plugins` config
3. Write in the line edit widget to add a filename pattern under `File extension readers`
4. Remove the text that was wrote.

### Traceback

<details>

```python-traceback
ValueError                                Traceback (most recent call last)
E:\Acer\Documentos\Quansight\Napari\napari\napari\_qt\widgets\qt_extension2reader.py in _filter_compatible_readers(self=<napari._vendor.qt_json_builder.qt_jsonschema_form.widgets.Extension2ReaderWidget object>, new_pattern='')
    171         readers = self._npe2_readers.copy()
    172         to_delete = []
--> 173         compatible_readers = get_potential_readers(new_pattern)
        compatible_readers = undefined
        global get_potential_readers = <function get_potential_readers at 0x000002B4CF38A320>
        new_pattern = ''
    174         for plugin_name in readers:
    175             if plugin_name not in compatible_readers:

E:\Acer\Documentos\Quansight\Napari\napari\napari\plugins\utils.py in get_potential_readers(filename='')
    132     # lower case file extension
    133     ext = str(Path(filename).suffix).lower()
--> 134     filename = str(Path(filename).with_suffix(ext))
        filename = ''
        global str = undefined
        global Path = <class 'pathlib.Path'>
        filename.with_suffix = undefined
        ext = ''
    135     for impl in hook_caller.get_hookimpls():
    136         reader = hook_caller._call_plugin(impl.plugin_name, path=filename)

~\anaconda3\envs\napari-dev\lib\pathlib.py in with_suffix(self=WindowsPath('.'), suffix='')
    780         name = self.name
    781         if not name:
--> 782             raise ValueError("%r has an empty name" % (self,))
        global ValueError = undefined
        self = WindowsPath('.')
    783         old_suffix = self.suffix
    784         if not old_suffix:

ValueError: WindowsPath('.') has an empty name
```

</details>

## Expected behavior

No error/traceback to happen

## Environment

<details>

```
napari: 0.5.0a2.dev231+g7689d09a
Platform: Windows-10-10.0.19045-SP0
Python: 3.10.9 | packaged by conda-forge | (main, Jan 11 2023, 15:15:40) [MSC v.1916 64 bit (AMD64)]
Qt: 5.15.2
PyQt5: 5.15.9
NumPy: 1.24.3
SciPy: 1.9.3
Dask: 2022.10.2
VisPy: 0.12.1
magicgui: 0.6.0
superqt: unknown
in-n-out: 0.1.5
app-model: 0.1.4
npe2: 0.7.0

OpenGL:
- GL version: 4.5.0 - Build 25.20.100.6444
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1366x768, scale 1.0
- screen 2: resolution 1600x900, scale 1.0

Settings path:
- C:\Users\dalth\AppData\Local\napari\napari-dev_4454fdbf170edefe23827a4bf512389a6806af69\settings.yaml
```

</details>

## Additional context

Issue found while checking #5977

### PR DESCRIPTION
**Pull request by Czaki**
# Fixes/Closes

Closes #6106

# Description

## Type of change
- [x] Bug-fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: napari/_qt/widgets/qt_extension2reader.py [hunk 1]
```diff
@@ -170,16 +170,17 @@ def _filter_compatible_readers(self, new_pattern):
...
+        else:
+            compatible_readers = {}
         for plugin_name in readers:
             if plugin_name not in compatible_readers:
                 to_delete.append(plugin_name)
 
         for reader in to_delete:
             del readers[reader]
 
-        readers.update(self._npe1_readers)
```
Comment (Czaki) [inline napari/_qt/widgets/qt_extension2reader.py:181]: npe1 readers are ignored by `self._add_reader_choice`, but impact if condition.

**Comment by DragaDoncila** [comment]
@Czaki I just had a look at this - sorry for the delay! I'm getting the same error on Mac, but this PR doesn't fix it - I get `ValueError: PosixPath('.') has an empty name` on both main and this branch? Haven't tested on Windows yet but just thought I should flag

File: napari/_qt/widgets/qt_extension2reader.py (hunk 1 above)
Comment (DragaDoncila) [reply napari/_qt/widgets/qt_extension2reader.py:181]: not sure what's given you that impression? On `main` with an npe1 reader installed I see the reader both in the initial list, and after filtering with an extension. In this branch, I no longer see the npe1 reader after filtering - so I think we do need this line.

**Comment by Czaki** [comment]
@DragaDoncila Could You recheck it and paste stacktrace if it really happens again? I could not reproduce bug on macos.

**Comment by DragaDoncila** [comment]
@Czaki I've tried this in a new environment and the error does still occur for me. Here is the stack trace:

```python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
File ~/CZI/napari/napari/_qt/widgets/qt_extension2reader.py:174, in Extension2ReaderTable._filter_compatible_readers(self=<napari._vendor.qt_json_builder.qt_jsonschema_form.widgets.Extension2ReaderWidget object>, new_pattern='.')
    172 to_delete = []
    173 if new_pattern:
--> 174     compatible_readers = get_potential_readers(new_pattern)
        new_pattern = '.'
    175 else:
    176     compatible_readers = {}

File ~/CZI/napari/napari/plugins/utils.py:134, in get_potential_readers(filename='.')
    132 # lower case file extension
    133 ext = str(Path(filename).suffix).lower()
--> 134 filename = str(Path(filename).with_suffix(ext))
        ext = ''
        filename = '.'
    135 for impl in hook_caller.get_hookimpls():
    136     reader = hook_caller._call_plugin(impl.plugin_name, path=filename)

File /opt/miniconda3/envs/napari/lib/python3.9/pathlib.py:903, in PurePath.with_suffix(self=PosixPath('.'), suffix='')
    901 name = self.name
    902 if not name:
--> 903     raise ValueError("%r has an empty name" % (self,))
        self = PosixPath('.')
        (self,) = (PosixPath('.'),)
    904 old_suffix = self.suffix
    905 if not old_suffix:

ValueError: PosixPath('.') has an empty name
```

It fires as soon as I type `.` into the text edit in the preferences dialog. Here is my `napari --info`:

```
napari: 0.5.0a2.dev331+gf3468e91
Platform: macOS-10.16-x86_64-i386-64bit
System: MacOS 13.5.2
Python: 3.9.17 (main, Jul  5 2023, 16:17:03)  [Clang 14.0.6 ]
Qt: 5.15.2
PyQt5: 5.15.9
NumPy: 1.25.2
SciPy: 1.11.2
Dask: 2023.9.0
VisPy: 0.13.0
magicgui: 0.7.3
superqt: 0.5.4
in-n-out: 0.1.8
app-model: 0.2.1
npe2: 0.7.2

OpenGL:
  - GL version:  2.1 INTEL-20.6.4
  - MAX_TEXTURE_SIZE: 16384

Screens:
  - screen 1: resolution 1440x900, scale 2.0

Settings path:
  - /Users/ddoncilapop/Library/Application Support/napari/napari_ea75b2750d5fa4e858db0ba9359fdfaa5175c19e/settings.yaml
Plugins:
  - napari: 0.5.0a2.dev331+gf3468e91 (77 contributions)
  - napari-console: 0.0.8 (0 contributions)
  - napari-svg: 0.1.10 (2 contributions)
```

**Comment by psobolewskiPhD** [comment]
I can confirm the same issue as @DragaDoncila on my mac.

The instructions are a bit weird in that panel?
they say `Enter a pattern ... e.g. "*.tif" ...`
I tired using the quotes--nope.
But! the `*` appears to be the key, as it prevents the error?

**Comment by DragaDoncila** [comment]
@psobolewskiPhD ugh yeah, I hated writing those instructions 😆 . The quotes were there to show that that is what you're typing exactly, but I can see how in coding land it would imply you should also type the quotes. Maybe we should remove?

The `*` is usually necessary, because you're **usually** wanting to match a pattern rather than a specific file path. So in most cases you'll need the `*`, but not in all (if you are trying to set a preference for a specific folder or something), and it doesn't have to be at the start - you could have `my-folder/*.tif` for example, if you wanted only TIFF files in `my-folder` to have the pattern associated. 

So yeah, it's super annoying and I don't know what the solution is besides just adding more examples to the instructions? But they're already long as it is... Maybe we should have just super basic instructions there and link out to a more detailed docs page? Stuff like this could definitely go into a `Guide to opening files and folders in napari`, which I think maybe you've suggested before?

For the purposes of this PR - you probably never want to start your pattern with just a `.` because that would almost never match any real file path, but I think that doesn't mean it should error, so we should still fix this.

**Comment by psobolewskiPhD** [comment]
To my mind, "a pattern" that uses `*` is a very complex concept for a GUI user, it's sort of like regex? The simple solution would be `Enter the extension ...` and then accept things that start with `.` like `.png` without error and then have more advanced cases in the docs?

**Comment by DragaDoncila** [comment]
@psobolewskiPhD we do actually prepend the `*` if the user just types in e.g. `.png`, so that use case is covered, but I can see how you might want to just simplify the instructions to mention extension, and then maybe just have a line that says "For more advanced use cases, check out the docs" (and then write the docs 😆). Anyway, that's out of scope for this PR so we should probably open an issue!

File: napari/_qt/widgets/qt_extension2reader.py (hunk 1 above)
Comment (DragaDoncila) [reply napari/_qt/widgets/qt_extension2reader.py:181]: Dummy npe1 reader for testing purposes: https://github.com/DragaDoncila/npe1-read

**Comment by Czaki** [comment]
Ok. I just end with catching exceptions and processing such use case.

**Review by DragaDoncila — APPROVED** [review]
Pulled it down and works as advertised. I wondered for a sec whether we should prevent people from saving preferences for "empty" patterns like `.` or `/`, but I actually think it's not our role to validate the pattern in that way and I think it's extremely unlikely for a user to accidentally do this. Approving!

**Pull request merged**